### PR TITLE
Leverage the match expression

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | -
| Tickets       | -
| License       | MIT
| Doc PR        | -

Tentative to leverage the `match` expression when we can in order to keep on modernizing the code (after `array_is_list()`, `str_starts_with()`, `str_ends_with()` and `str_contains()`).

Firstly, let's see if the CI is green.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -1559,10 +1559,10 @@ public function getParameterBag(): ParameterBagInterface
         if ($dynamicPhp) {
             $loadedDynamicParameters = $this->exportParameters(array_combine(array_keys($dynamicPhp), array_fill(0, \count($dynamicPhp), false)), '', 8);
             $getDynamicParameter = <<<'EOF'
-        switch ($name) {
+        $value = match ($name) {
```
Comment (fancyweb) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:1562]: @nicolas-grekas This change is not really minor I guess, should I submit it as a feature in another PR?

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php [hunk 2]
```diff
@@ -297,19 +297,12 @@ private function loadFromFile(ContainerBuilder $container, $file, $format)
...
-                $loader = new YamlFileLoader($container, $locator);
-                break;
-            default:
-                throw new \InvalidArgumentException(sprintf('Unsupported format: "%s"', $format));
-        }
+        $loader = match ($format) {
+            'php' => new PhpFileLoader($container, $locator),
+            'xml' => new XmlFileLoader($container, $locator),
+            'yml' => new YamlFileLoader($container, $locator),
+            default => throw new \InvalidArgumentException(sprintf('Unsupported format: "%s"', $format)),
```
Comment (derrabus) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: You could omit the `default` case here. `match` will raise an error if no case matches. And this isn't production code, so we don't need to bother about beautiful exception messages.

File: src/Symfony/Component/Config/Definition/Dumper/XmlReferenceDumper.php [hunk 3]
```diff
@@ -100,27 +100,13 @@ private function writeNode(NodeInterface $node, int $depth = 0, bool $root = fal
...
-                                break;
-
-                            default:
-                                $prototypeValue = 'value';
-                        }
+                        $prototypeValue = match (\get_class($prototype)) {
+                            'Symfony\Component\Config\Definition\ScalarNode' => 'scalar value',
+                            'Symfony\Component\Config\Definition\FloatNode', 'Symfony\Component\Config\Definition\IntegerNode' => 'numeric value',
+                            'Symfony\Component\Config\Definition\BooleanNode' => 'true|false',
+                            'Symfony\Component\Config\Definition\EnumNode' => implode('|', array_map('json_encode', $prototype->getValues())),
```
Comment (derrabus) [inline src/Symfony/Component/Config/Definition/Dumper/XmlReferenceDumper.php]: While we're at it, can we switch the FQCN strings to `::class`?

File: src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php [hunk 4]
```diff
@@ -166,14 +166,10 @@ public function guessTypeForConstraint(Constraint $constraint): ?TypeGuess
...
-        switch (\get_class($constraint)) {
-            case 'Symfony\Component\Validator\Constraints\NotNull':
-            case 'Symfony\Component\Validator\Constraints\NotBlank':
-            case 'Symfony\Component\Validator\Constraints\IsTrue':
-                return new ValueGuess(true, Guess::HIGH_CONFIDENCE);
-        }
-
-        return null;
+        return match (\get_class($constraint)) {
+            'Symfony\Component\Validator\Constraints\NotNull', 'Symfony\Component\Validator\Constraints\NotBlank', 'Symfony\Component\Validator\Constraints\IsTrue' => new ValueGuess(true, Guess::HIGH_CONFIDENCE),
```
Comment (derrabus) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php]: Let's switch to `::class` here as well.

**Review by derrabus — COMMENTED** [review]
Very nice!

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 5]
```diff
@@ -51,44 +51,23 @@ public function guessType(string $class, string $property): ?TypeGuess
...
-            case Types::STRING:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::MEDIUM_CONFIDENCE);
-            case Types::TEXT:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextareaType', [], Guess::MEDIUM_CONFIDENCE);
-            default:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::LOW_CONFIDENCE);
-        }
+        return match ($metadata->getTypeOfField($property)) {
+            Types::ARRAY, Types::SIMPLE_ARRAY => new TypeGuess('Symfony\Component\Form\Extension\Core\Type\CollectionType', [], Guess::MEDIUM_CONFIDENCE),
+            Types::BOOLEAN => new TypeGuess('Symfony\Component\Form\Extension\Core\Type\CheckboxType', [], Guess::HIGH_CONFIDENCE),
```
Comment (GromNaN) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: We could use the `::class` syntax with a namespace import?

**Comment by fancyweb** [comment]
I'm totally in favor of switching to FQCN but it needs to be done globally in another PR I guess. That's unrelated to the match expression for me.

**Comment by nicolas-grekas** [comment]
We prefer changing lines once in a single PR, because that helps when running git blame later on.

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 6]
```diff
@@ -51,44 +51,23 @@ public function guessType(string $class, string $property): ?TypeGuess
...
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\IntegerType', [], Guess::MEDIUM_CONFIDENCE);
-            case Types::STRING:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::MEDIUM_CONFIDENCE);
-            case Types::TEXT:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextareaType', [], Guess::MEDIUM_CONFIDENCE);
-            default:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::LOW_CONFIDENCE);
-        }
+        return match ($metadata->getTypeOfField($property)) {
+            Types::ARRAY, Types::SIMPLE_ARRAY => new TypeGuess('Symfony\Component\Form\Extension\Core\Type\CollectionType', [], Guess::MEDIUM_CONFIDENCE),
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: lines with multiple values should be split on many lines

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]:
```
Types::ARRAY,
Types::SIMPLE_ARRAY,
    => ...
```

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:1562]: I'd keep it here, it's the very same topic

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 7]
```diff
@@ -51,44 +61,44 @@ public function guessType(string $class, string $property): ?TypeGuess
...
-            case Types::SMALLINT:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\IntegerType', [], Guess::MEDIUM_CONFIDENCE);
-            case Types::STRING:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::MEDIUM_CONFIDENCE);
-            case Types::TEXT:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextareaType', [], Guess::MEDIUM_CONFIDENCE);
-            default:
-                return new TypeGuess('Symfony\Component\Form\Extension\Core\Type\TextType', [], Guess::LOW_CONFIDENCE);
-        }
+        return match ($metadata->getTypeOfField($property)) {
```
Comment (fancyweb) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php:64]: @nicolas-grekas I modified just this match with your suggestion, would that code style works for you?

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php:64]: Now yes!

**Comment by derrabus** [comment]
Looks like PHP CS Fixer does not like the code style proposed by @nicolas-grekas (yet).

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 8]
```diff
@@ -21,6 +21,16 @@
...
+use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\Extension\Core\Type\DateTimeType;
+use Symfony\Component\Form\Extension\Core\Type\DateIntervalType;
+use Symfony\Component\Form\Extension\Core\Type\DateType;
+use Symfony\Component\Form\Extension\Core\Type\TimeType;
+use Symfony\Component\Form\Extension\Core\Type\NumberType;
+use Symfony\Component\Form\Extension\Core\Type\IntegerType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\Extension\Core\Type\TextareaType;
```
Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php:27]: CS: The `use` statements should be sorted.

**Comment by nicolas-grekas** [comment]
> Looks like PHP CS Fixer does not like the code style proposed by @nicolas-grekas (yet).

I'm fine adopting php-cs-fixer suggestion then (after removing the trailing coma)

File: src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php [hunk 9]
```diff
@@ -83,21 +83,15 @@ private function phpize(string $value): mixed
...
-            case \defined($value):
-                return \constant($value);
-            case 'yes' === $lowercaseValue || 'on' === $lowercaseValue:
-                return true;
-            case 'no' === $lowercaseValue || 'off' === $lowercaseValue || 'none' === $lowercaseValue:
-                return false;
-            case isset($value[1]) && (
+        return match (true) {
+            \defined($value) => \constant($value),
+            'yes' === $lowercaseValue || 'on' === $lowercaseValue => true,
```
Comment (jderusse) [inline src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php]:
```suggestion
            'yes' === $lowercaseValue, 'on' === $lowercaseValue => true,
```

?

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @fancyweb.
